### ISSUE TITLE
Live Migration. SSH session to migrating VM can be lost.

### ISSUE THREAD

**Issue by DaveLangridge** [Metaswitch/calico#188]
I had an ssh session to a VM from host 2. Live migrating to host 1 and back. Migrating from host 1 to 2, ssh session stayed up with gap in pings to another VM ~2s. Migrating from host 2 to host 1, I get 'broken pipe' for the ssh connection. The VM has migrated OK, pings to the migrating VM show a gap of ~2s.

**Comment by ed-harrison** [comment]
@Lukasa Can you have a look at this to double check whether this is something we should be worrying about or prioritising?

**Comment by Lukasa** [comment]
EPIPE is usually raised because the connection has been closed on the far end, but I wouldn't be at all surprised to find it gets returned if an ICMP Destination Host Unreachable message gets received. This can happen when moving VMs in a small window when there is no route available for that VM, and the window is larger if the VM is moving *away* from the host that has the SSH session open.

I think this is relatively non-urgent: I'd only consider it problematic if we can reproduce this with traffic moving between VMs or from outside the data center.

**Comment by fasaxc** [comment]
I don't think we do anything special for migrations right now, we treat them as a removal and then and add of a port.  I feel like, to handle migration robustly without dropping connections, we need to:

* keep the route on the old host until the new host publishes its route to the guest
* blackhole traffic for the guest at the old host until the traffic all moves to the new
* remove the route on the old host
* stop blackholing traffic at old guest.

It's probably good enough to have a 60s overlap rather than trying to detect convergence.

Without that, there's a window where there's no route to the guest on either node so downstream routers will reply with no-route-to-host.

**Comment by eepyaich** [comment]
Boiling this down to the customer perspective, it sounds like there is a window condition where live migration will fail (i.e. if the definition of "live" is that my SSH session didn't drop). @fasaxc @DaveLangridge is this a reasonable summary?

**Comment by DaveLangridge** [comment]
Yes. The migration works, but the SSH sometimes drops and you need to reconnect.

**Comment by fasaxc** [comment]
We could use BGP to make this go very smoothly, assuming we can spot when something is being migrated:

* host that initially contains workload starts advertising route with a community that indicates that the VM is being migrated (we time out this route after, say 60s)
* new host starts advertising route to the VM without community (i.e. just normal advertisement)
* network is configured to prefer normal routes to routes with the migration community attached.

There's always a route to the address of the workload and the best route is the new one as soon as it comes up.

**Comment by neiljerram** [comment]
Note: this is also being tracked as a (possible) networking-calico issue at https://bugs.launchpad.net/networking-calico/+bug/1628960.
